skrifa/src/variation.rs

```diff
@@ -0,0 +1,539 @@
+//! Axes of variation in a variable font.
+
+use read_fonts::{
+    tables::avar::Avar,
+    tables::fvar::{self, Fvar},
+    types::{Fixed, Tag},
+    TableProvider,
+};
+
+use crate::{
+    instance::{Location, NormalizedCoord},
+    setting::VariationSetting,
+    string::StringId,
+    util::SmallArray,
+};
+
+/// Axis of variation in a variable font.
+///
+/// In variable fonts, an axis usually refers to a single aspect of a
+/// typeface's design that can be altered by the user.
+///
+/// See <https://fonts.google.com/knowledge/glossary/axis_in_variable_fonts>
+#[derive(Clone)]
+pub struct Axis {
+    index: usize,
+    record: fvar::VariationAxisRecord,
+}
+
+impl Axis {
+    /// Returns the tag that identifies the axis.
+    pub fn tag(&self) -> Tag {
+        self.record.axis_tag()
+    }
+
+    /// Returns the index of the axis in its owning collection.
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// Returns the localized string identifier for the name of the axis.
+    pub fn name_id(&self) -> StringId {
+        self.record.axis_name_id()
+    }
+
+    /// Returns true if the axis should be hidden in user interfaces.
+    pub fn is_hidden(&self) -> bool {
+        const AXIS_HIDDEN_FLAG: u16 = 0x1;
+        self.record.flags() & AXIS_HIDDEN_FLAG != 0
+    }
+
+    /// Returns the minimum value of the axis.
+    pub fn min_value(&self) -> f32 {
+        self.record.min_value().to_f64() as _
+    }
+
+    /// Returns the default value of the axis.
+    pub fn default_value(&self) -> f32 {
+        self.record.default_value().to_f64() as _
+    }
+
+    /// Returns the maximum value of the axis.
+    pub fn max_value(&self) -> f32 {
+        self.record.max_value().to_f64() as _
+    }
+
+    /// Returns a normalized coordinate for the given user coordinate. The value will be
+    /// clamped to the range specified by the minimum and maximum values.
+    ///
+    /// This does not apply any axis variation remapping.
+    pub fn normalize(&self, coord: f32) -> NormalizedCoord {
+        self.record
+            .normalize(Fixed::from_f64(coord as _))
+            .to_f2dot14()
+    }
+}
+
+/// Collection of axes in a variable font.
+///
+/// See the [`Axis`] type for more detail.
+#[derive(Clone)]
+pub struct AxisCollection<'a> {
+    fvar: Option<Fvar<'a>>,
+    avar: Option<Avar<'a>>,
```

Comment:
Do we need that / does avar need to be public in `AxisCollection` as client API?

Reply:
I’m not sure I understand the question. It’s an implementation detail here for computing the final location in normalized space. My intention is that users shouldn’t have to deal with axis remapping explicitly at this level.

Reply:
So, `AxisCollection` is public, but the fields are not? Is that right?

Reply:
Correct! In Rust, all fields are private unless otherwise specified.